openingd/dualopend.c

```diff
@@ -2349,6 +2346,178 @@ static void opener_start(struct state *state, u8 *msg)
+	 * transaction feerate, i.e. if the initial `feerate_per_kw_funding`
+	 * was 512 satoshis per kiloweight, `fee_step` 1 is
+	 * 512 + 512 / 4 or 640 sat/kw, `fee_step` 2
+	 * is 640 + 640 / 4 or 800 sat/kw.
+	 */
+	for (size_t i = fee_step; i > 0; i--)
+		feerate = feerate + feerate / 4;
+
+	/* It's possible they sent us a 'bad' feerate step,
+	 * i.e. less than the last one. */
+	if (feerate < last_feerate)
+		return false;
+
+	tx_state->feerate_per_kw_funding = feerate;
+	return true;
+}
+
+static void rbf_start(struct state *state, const u8 *rbf_msg)
+{
+	struct channel_id cid;
+	struct tx_state *tx_state;
+	char *err_reason;
+	struct amount_sat total;
+	enum dualopend_wire msg_type;
+	u8 fee_step, *msg;
+
+	/* We need a new tx_state! */
+	tx_state = new_tx_state(state);
+
+	if (!fromwire_init_rbf(rbf_msg, &cid,
+			       &tx_state->opener_funding,
+			       &tx_state->tx_locktime,
+			       &fee_step))
+		peer_failed_err(state->pps, &state->channel_id,
+				"Parsing init_rbf %s",
+				tal_hex(tmpctx, rbf_msg));
+
+	/* Is this the correct channel? */
+	check_channel_id(state, &cid, &state->channel_id);
+	peer_billboard(false, "channel rbf: init received from peer");
+
+	/* Have you sent us everything we need yet ? */
+	if (!state->last_tx_state->remote_funding_sigs_rcvd)
+		rbf_failed(state, "Last funding attempt not complete:"
+			   " missing your funding tx_sigs");
+
+	/* FIXME: should we check for currently in progress? */
+
+	/* Copy over the channel config info -- everything except
+	 * the reserve will be the same */
+	tx_state->localconf = state->tx_state->localconf;
+	tx_state->remoteconf = state->tx_state->remoteconf;
+
+	if (!update_feerate(tx_state,
+			    state->feerate_per_kw_funding,
+			    state->last_tx_state->feerate_per_kw_funding,
+			    fee_step)) {
+		rbf_failed(state, "Fee step not greater than last."
+			   " Fee step %d, last feerate %d",
+			   fee_step,
+			   state->last_tx_state->feerate_per_kw_funding);
+		return;
+	}
+
+	/* We've copied everything we need over now. Promote
+	 * tx_state to last_tx_state */
+	tal_free(state->last_tx_state);
+	state->last_tx_state = state->tx_state;
+	state->tx_state = tx_state;
+
+	/* We ask master if this is ok */
+	msg = towire_dualopend_got_rbf_offer(NULL,
+					     &state->channel_id,
+					     tx_state->opener_funding,
+					     tx_state->feerate_per_kw_funding,
+					     tx_state->tx_locktime);
+
+	wire_sync_write(REQ_FD, take(msg));
+	msg = wire_sync_read(tmpctx, REQ_FD);
+
+	if ((msg_type = fromwire_peektype(msg)) == WIRE_DUALOPEND_FAIL) {
+		if (!fromwire_dualopend_fail(msg, msg, &err_reason))
+			master_badmsg(msg_type, msg);
+		rbf_failed(state, "%s", err_reason);
+	}
+
+	if (!fromwire_dualopend_got_rbf_offer_reply(state, msg,
+						    &tx_state->accepter_funding,
+						    &tx_state->psbt))
+		master_badmsg(WIRE_DUALOPEND_GOT_RBF_OFFER_REPLY, msg);
+
+	if (!tx_state->psbt)
+		tx_state->psbt = create_psbt(tx_state, 0, 0,
+					     tx_state->tx_locktime);
```

Comment:
I think this should be assigned unconditionally?  Aren't we otherwise just reusing the last one (directly; we don't copy it!).